primitives/src/rescue/sponge.rs

```diff
@@ -0,0 +1,420 @@
+//! This file contains the APIs wrappers for ark-sponge
+
+use ark_ff::{BigInteger, PrimeField};
+use ark_sponge::{
+    Absorb, CryptographicSponge, FieldBasedCryptographicSponge, FieldElementSize, SpongeExt,
+};
+use ark_std::{string::ToString, vec, vec::Vec};
+use jf_utils::{field_switching, pad_with_zeros};
+use num_bigint::BigUint;
+
+use super::{errors::RescueError, Permutation, RescueParameter, RescueVector, RATE, STATE_SIZE};
+
+#[derive(Clone, Default)]
+/// A rescue hash function consists of a permutation function and
+/// an internal state.
+pub struct RescueSponge<F: RescueParameter, const ABSORB_RATE: usize> {
+    pub(crate) state: RescueVector<F>,
+}
+
+/// CRHF
+pub type RescueSpongeCRHF<F> = RescueSponge<F, 3>;
+/// PRF
+pub type RescueSpongePRF<F> = RescueSponge<F, 4>;
+
+impl<F: RescueParameter, const CHUNK_SIZE: usize> RescueSponge<F, CHUNK_SIZE> {
+    /// Sponge hashing based on rescue permutation for Bls12_381 scalar field
+    /// for RATE 3 and CAPACITY 1. It allows unrestricted variable length
+    /// input and number of output elements
+    pub fn sponge_with_padding(input: &[F], num_outputs: usize) -> Vec<F> {
+        // Pad input as follows: append a One, then pad with 0 until length is multiple
+        // of RATE
+        let mut padded = input.to_vec();
+        padded.push(F::one());
+        pad_with_zeros(&mut padded, RATE);
```

Comment:
Should here be `CHUNK_SIZE` or `RATE`?

Reply:
I'm now enforcing these to be the same in this implementation, see comment below.